lib/SILGen/SILGenRequests.cpp

```diff
@@ -56,6 +56,49 @@ evaluator::DependencySource SILGenSourceFileRequest::readDependencySource(
   };
 }
 
+ArrayRef<FileUnit *> SILGenDescriptor::getFiles() const {
+  if (auto *mod = context.dyn_cast<ModuleDecl *>())
+    return mod->getFiles();
+
+  // For a single file, we can form an ArrayRef that points at its storage in
+  // the union.
+  return llvm::makeArrayRef(*context.getAddrOfPtr1());
+}
+
+bool SILGenDescriptor::isWholeModule() const {
+  return context.is<ModuleDecl *>();
+}
+
+SourceFile *SILGenDescriptor::getSourceFileToParse() const {
+#ifndef NDEBUG
+  auto sfCount = llvm::count_if(getFiles(), [](FileUnit *file) {
+    return isa<SourceFile>(file);
+  });
+  auto silFileCount = llvm::count_if(getFiles(), [](FileUnit *file) {
+    auto *SF = dyn_cast<SourceFile>(file);
+    return SF && SF->Kind == SourceFileKind::SIL;
+  });
+  assert(silFileCount == 0 || (silFileCount == 1 && sfCount == 1) &&
+         "Cannot currently mix a .sil file with other SourceFiles");
+#endif
+
+  for (auto *file : getFiles()) {
+    // Skip other kinds of files.
+    auto *SF = dyn_cast<SourceFile>(file);
+    if (!SF)
+      continue;
+
+    // Given the above precondition that a .sil file isn't mixed with other
+    // SourceFiles, we can return a SIL file if we have it, or return nullptr.
+    if (SF->Kind == SourceFileKind::SIL) {
```

Comment:
Would it be difficult to allow mixing SIL with swift source in the same module?

Reply:
(This could allow writing parts of the stdlib in SIL, which was an idea @jckarter had at some point...)

Reply:
@slavapestov I don't _think_ it would be too difficult, we'd just need to make sure that we don't mix canonical parsed SIL with raw generated SIL.